qmetaobject/src/qtdeclarative.rs

```diff
@@ -201,6 +202,106 @@ impl Default for QQuickView {
     }
 }
 
+#[repr(u32)]
+#[derive(Clone, Copy)]
+pub enum CompilationMode {
+    PreferSynchronous,
+    Asynchronous,
+}
+
+#[repr(u32)]
+#[derive(Clone, Copy)]
+pub enum ComponentStatus {
+    Null,
+    Ready,
+    Loading,
+    Error
+}
+
+cpp!{{
+    struct QQmlComponentHolder {
+        std::unique_ptr<QQmlComponent> component;
+
+        QQmlComponentHolder(QQmlEngine *e) : component(new QQmlComponent(e)) {}
+    };
+}}
+
+cpp_class!(
+    /// Wrapper for QQmlComponent
+    pub unsafe struct QmlComponent as "QQmlComponentHolder"
+);
+
+impl QmlComponent {
+    /// Create a QmlComponent using the QmlEngine.
+    pub fn new(engine: &QmlEngine) -> QmlComponent {
+        unsafe {
+            cpp!([engine as "QmlEngineHolder*"] -> QmlComponent as "QQmlComponentHolder" {
+                return QQmlComponentHolder(engine->engine.get());
+            })
+        }
+    }
+
+    pub fn get_cpp_object(&self) -> *mut c_void {
+        unsafe { cpp!([self as "QQmlComponentHolder*"] -> *mut c_void as "QQmlComponent*" {
+            return self->component.get();
+        })}
+    }
+
+    /// Performs QQmlComponent::loadUrl
+    pub fn load_url(&mut self, url: QUrl, compilation_mode: CompilationMode) {
+        unsafe { cpp!([self as "QQmlComponentHolder*", url as "QUrl", compilation_mode as "QQmlComponent::CompilationMode"]{
+            self->component->loadUrl(url, compilation_mode);
+        })}
+    }
+
+    /// Performs QQmlComponent::setData with a default url
+    pub fn set_data(&mut self, data: QByteArray) {
+        unsafe { cpp!([self as "QQmlComponentHolder*", data as "QByteArray"]{
+            self->component->setData(data, QUrl());
+        })}
+    }
+
+    /// Performs QQmlComponent::setData
+    pub fn set_data_as(&mut self, data: QByteArray, url: QUrl) {
+        unsafe { cpp!([self as "QQmlComponentHolder*", data as "QByteArray", url as "QUrl"]{
+            self->component->setData(data, url);
+        })}
+    }
+
+    /// Performs QQmlComponent::create
+    pub fn create(&mut self) -> *mut c_void {
+        unsafe { cpp!([self as "QQmlComponentHolder*"] -> *mut c_void as "QObject*" {
+            return self->component->create();
+        })}
+    }
+
+    /// See Qt documentation for QObject::destroyed
+    pub fn status_changed_signal() -> CppSignal<fn(status: ComponentStatus)> {
```

Comment:
Documentation needs fix.  
I also would like to see this used in a test, please.

Reply:
I've copied ```test_future``` as a template for implementing this test. In so doing I've noticed that both my test and ```test_future``` crash on macOS Catalina with Qt 5.13.1 and nightly Rust.